provider/formtype_test.go

```diff
@@ -0,0 +1,426 @@
+						Options:    %t,
+						OptionType: %q,
+						FormType:   %q,
+					}),
+				//`, "<expected_form_type>", check.options, check.optionType, check.formType)
+
+				probablyPassed := formTypeTest(t, fc)
+				if !probablyPassed {
+					t.Logf("To construct this test case:\n%s", tcText)
+				}
+			})
+
+		}
+	})
+}
+
+// ezconfig converts a formTypeCheck into a terraform config string.
+func ezconfig(paramName string, cfg formTypeCheck) (defaultValue string, tf string) {
+	options := cfg.customOptions
+	if cfg.options && len(cfg.customOptions) == 0 {
+		switch cfg.optionType {
+		case provider.OptionTypeString:
+			options = []string{"foo"}
+			defaultValue = "foo"
+		case provider.OptionTypeBoolean:
+			options = []string{"true", "false"}
+			defaultValue = "true"
+		case provider.OptionTypeNumber:
+			options = []string{"1"}
+			defaultValue = "1"
+		case provider.OptionTypeListString:
+			options = []string{`["red", "blue"]`}
+			defaultValue = `["red", "blue"]`
+		default:
+			panic(fmt.Sprintf("unknown option type %q when generating options", cfg.optionType))
+		}
+	}
+
+	if cfg.defValue == "" {
+		cfg.defValue = defaultValue
+	}
+
+	var body strings.Builder
+	if cfg.defValue != "" {
+		body.WriteString(fmt.Sprintf("default = %q\n", cfg.defValue))
+	}
+	if cfg.formType != "" {
+		body.WriteString(fmt.Sprintf("form_type = %q\n", cfg.formType))
+	}
+	if cfg.optionType != "" {
+		body.WriteString(fmt.Sprintf("type = %q\n", cfg.optionType))
+	}
+	if cfg.styling != nil {
+		body.WriteString(fmt.Sprintf("styling = %s\n", strconv.Quote(string(cfg.styling))))
+	}
+
+	for i, opt := range options {
+		body.WriteString("option {\n")
+		body.WriteString(fmt.Sprintf("name = \"val_%d\"\n", i))
+		body.WriteString(fmt.Sprintf("value = %q\n", opt))
+		body.WriteString("}\n")
+	}
+
+	return cfg.defValue, fmt.Sprintf(`
+				provider "coder" {
+				}
+				data "coder_parameter" "%s" {
+					name = "%s"
+					%s
+				}
+		`, paramName, paramName, body.String())
+}
+
+func formTypeTest(t *testing.T, c formTypeTestCase) bool {
+	t.Helper()
+	const paramName = "test_param"
+	// probablyPassed is just a guess used for logging. It's not important.
+	probablyPassed := true
+
+	def, tf := ezconfig(paramName, c.config)
+	checkFn := func(state *terraform.State) error {
+		require.Len(t, state.Modules, 1)
+		require.Len(t, state.Modules[0].Resources, 1)
+
+		key := strings.Join([]string{"data", "coder_parameter", paramName}, ".")
+		param := state.Modules[0].Resources[key]
+
+		probablyPassed = probablyPassed && assert.Equal(t, def, param.Primary.Attributes["default"], "default value")
+		probablyPassed = probablyPassed && assert.Equal(t, string(c.assert.FormType), param.Primary.Attributes["form_type"], "form_type")
+		probablyPassed = probablyPassed && assert.Equal(t, string(c.assert.Type), param.Primary.Attributes["type"], "type")
+		probablyPassed = probablyPassed && assert.JSONEq(t, string(c.assert.Styling), param.Primary.Attributes["styling"], "styling")
```

Comment:
Suggestion: rewrite the assertion chain and use a buffered channel to pass the result back.
```
resultCh <- assert.Equal(cond1) && assert.Equal(cond2) && ...
```

Reply:
If you do that, then the first failing `assert` stops the other assertions. 
I think I'd rather have each assertion dump it's failing output.